### ISSUE TITLE
need updated versioned checks on glib and GTK+2

### ISSUE THREAD

**Issue by nieder** [DeaDBeeF-Player/deadbeef#2630]
### Steps to reproduce the problem
Building against gtk2-2.18.9 and glib2-(yes, I know they're old) succeeds, but runtime fails due to not finding `g_get_monotonic_time` (from #2585) and `gtk_widget_get_mapped` (from 3cc0438). 

### What's going on? Describe the problem in as much detail as possible.

`g_get_monotonic_time` was first available with [glib-2.28](https://developer.gnome.org/glib/unstable/glib-Date-and-Time-Functions.html). `gtk_widget_get_mapped` was first available with [gtk-2.20](https://developer.gnome.org/gtk2/stable/GtkWidget.html#gtk-widget-get-mapped).

Currently, the gtk+2 test in configure only looks for gtk+2 >= 2.12 and the glib-2.0 check only looks for the presence of the pkg and doesn't do a versioned check.

### Information about the software:

Deadbeef version: HEAD
OS: macOS 10.14.6

**Comment by Alexey-Yakovenko** [comment]
Thanks for spotting the issue.
Although, it's not exactly the way you see it. The issue is that deadbeef should not be using g_get_monotonic_time at all, and is supposed to work with gtk2.16+

**Comment by Alexey-Yakovenko** [comment]
I applied a fix, but currently I don't have a way of validating it. Could you please give it a try?

**Comment by nieder** [comment]
If I directly replace all instances of `gtk_widget_get_mapped` in ddb_splitter.c with `GTK_WIDGET_MAPPED` rather than using a #define, then startup fails looking for `gtk_widget_style_attach`.

https://github.com/DeaDBeeF-Player/deadbeef/blob/e0f659dd56bc02b84000db1622148820de0f8ee8/plugins/gtkui/ddb_splitter.c#L645-L647

Swapping the above code (using the commented out lines instead of `gtk_widget_style_attach()` ) lets everything build, load, and play music.

**Comment by Alexey-Yakovenko** [comment]
Thanks. I realized I accidentally added the gtk_widget_get_mapped macro to the pltbrowser header, instead of gtkui. Fixing this now.

The other problem with gtk_widget_style_attach -- looking into it.. your solution seems a bit suspicious to me.

**Comment by Alexey-Yakovenko** [comment]
I applied the fix, according to your findings. Also tested with both GTK2 and GTK3 with split view on linux (20.04), and it seems to work the same as before.

**Comment by nieder** [comment]
I'm still getting a load failure for `_gtk_widget_get_mapped`, even with `plugins/gtkui/support.h` patched in 6d3b112. I had to edit `ddb_splitter.c` directly by either adding the #define there, or by directly changing `gtk_widget_get_mapped` with GTK_WIDGET_MAPPED (everything else about the glib/gtk symbols looks fixed.

I thought it might be something being included before `"support.h"`, but this is my compiler line and it looks OK:

>make[3]: Entering directory '/src/deadbeef/plugins/gtkui'
/bin/sh ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../..   -I/sw/include -I/opt/X11/include -I/sw/lib/ffmpeg-4.0/include -std=c99 -I/sw/include/gtk-2.0 -I/sw/lib/gtk-2.0/include -I/sw/include/gtk-2.0 -I/sw/include/pango-1.0 -I/sw/include/atk-1.0 -I/sw/include/cairo -I/sw/include/pango-1.0 -I/sw/include/freetype2 -I/sw/include -I/sw/include/glib-2.0 -I/sw/lib/glib-2.0/include -I/opt/X11/include  -I/sw/include -DDDB_WARN_DEPRECATED=1 -g -O2  -D_GNU_SOURCE -D__EXTENSIONS__  -DLIBDIR=\"/tmp/prefix/lib\" -DPREFIX=\"/tmp/prefix\" -DDOCDIR=\"/tmp/prefix/share/doc/deadbeef\" -DDDB_WARN_DEPRECATED=1 -c -o ddb_gui_GTK2_la-ddb_splitter.lo `test -f 'ddb_splitter.c' || echo './'`ddb_splitter.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../.. -I/sw/include -I/opt/X11/include -I/sw/lib/ffmpeg-4.0/include -std=c99 -I/sw/include/gtk-2.0 -I/sw/lib/gtk-2.0/include -I/sw/include/gtk-2.0 -I/sw/include/pango-1.0 -I/sw/include/atk-1.0 -I/sw/include/cairo -I/sw/include/pango-1.0 -I/sw/include/freetype2 -I/sw/include -I/sw/include/glib-2.0 -I/sw/lib/glib-2.0/include -I/opt/X11/include -I/sw/include -DDDB_WARN_DEPRECATED=1 -g -O2 -D_GNU_SOURCE -D__EXTENSIONS__ -DLIBDIR=\"/tmp/prefix/lib\" -DPREFIX=\"/tmp/prefix\" -DDOCDIR=\"/tmp/prefix/share/doc/deadbeef\" -DDDB_WARN_DEPRECATED=1 -c ddb_splitter.c  -fno-common -DPIC -o .libs/ddb_gui_GTK2_la-ddb_splitter.o

**Comment by Alexey-Yakovenko** [comment]
Are you sure you're building the master branch?
I'm not seeing references to the _gtk_widget_get_mapped in the ddb_gui_GTK2.so:

```
strings /Users/waker/Downloads/deadbeef-devel/plugins/ddb_gui_GTK2.so | grep _gtk_widget_get_mapped
```

**Comment by Alexey-Yakovenko** [comment]
Also are you sure that the version of GTK that you are using on the build machine is below 2.18?

**Comment by Alexey-Yakovenko** [comment]
I just double checked the command line and warnings when compiling with GTK2.16, and there are no warnings about missing symbols:

```
/bin/bash ../../libtool  --tag=CC   --mode=compile /usr/src/deadbeef/external/apbuild/apgcc -DHAVE_CONFIG_H -I. -I../..    -std=c99 -I/usr/src/deadbeef/static-deps/lib-x86-64/gtk-2.16.0/include/gtk-2.0 -I/usr/src/deadbeef/static-deps/lib-x86-64/gtk-2.16.0/lib/gtk-2.0/include -I/usr/src/deadbeef/static-deps/lib-x86-64/gtk-2.16.0/include/atk-1.0 -I/usr/src/deadbeef/static-deps/lib-x86-64/gtk-2.16.0/include/cairo -I/usr/src/deadbeef/static-deps/lib-x86-64/gtk-2.16.0/include/pango-1.0 -I/usr/src/deadbeef/static-deps/lib-x86-64/gtk-2.16.0/include -I/usr/src/deadbeef/static-deps/lib-x86-64/gtk-2.16.0/include/glib-2.0 -I/usr/src/deadbeef/static-deps/lib-x86-64/gtk-2.16.0/lib/glib-2.0/include  -I../../static-deps/lib-x86-64/include -DDDB_WARN_DEPRECATED=1 -m64 -I/usr/src/deadbeef/static-deps/lib-x86-64/include -O3 -D_FORTIFY_SOURCE=0  -D_GNU_SOURCE -D__EXTENSIONS__  -DLIBDIR=\"/opt/deadbeef/lib\" -DPREFIX=\"/opt/deadbeef\" -DDOCDIR=\"/opt/deadbeef/share/doc/deadbeef\" -DDDB_WARN_DEPRECATED=1 -c -o ddb_gui_GTK2_la-ddb_splitter.lo `test -f 'ddb_splitter.c' || echo './'`ddb_splitter.c
```

Therefore, I'm assuming that everything works correctly, and you are either not building the master branch, or your GTK headers are >=2.18.

**Comment by Alexey-Yakovenko** [comment]
BTW If I may ask -- why would you even want to build GTK2 UI on mac?

**Comment by nieder** [comment]
The above command is definitely from git master branch.
When I try to run an unpatched build, I get this error when I try to run a compiled deadbeef (compilation goes OK):
```
loading plugin /tmp/prefix/lib/deadbeef/ddb_gui_GTK2.so
dlopen error: dlopen(/tmp/prefix/lib/deadbeef/ddb_gui_GTK2.so, 2): Symbol not found: _gtk_widget_get_mapped
  Referenced from: /tmp/prefix/lib/deadbeef/ddb_gui_GTK2.so
  Expected in: flat namespace
 in /tmp/prefix/lib/deadbeef/ddb_gui_GTK2.so
trying /tmp/prefix/lib/deadbeef/ddb_gui_GTK2.fallback.so...
the plugin not found or failed to load
```

And this is the compiled file:
```
$ strings plugins/gtkui/.libs/ddb_gui_GTK2.so | grep gtk_widget_get_mapped
$ nm -o plugins/gtkui/.libs/ddb_gui_GTK2.so | grep gtk_widget_get_mapped
plugins/gtkui/.libs/ddb_gui_GTK2.so:                  U _gtk_widget_get_mapped
```

**Comment by nieder** [comment]
I'm building GTK+2 on a mac because sometimes it's more convenient to have a non .app bundle: such as an ssh remote session or just command line control.

**Comment by Alexey-Yakovenko** [comment]
you can control deadbeef from command line / over ssh even if it's an app bundle.

anyway, whatever you prefer.

The fix that I applied has this: `#if !GTK_CHECK_VERSION(2,18,0)`
Can you confirm that the GTK headers you're using are not 2.18+?

**Comment by Alexey-Yakovenko** [comment]
It seems that the error is caused by this combination of factors:

1. You build gtkui with GTK2 >= 2.18
2. You run gtkui with GTK2 < 2.18

This would not work. If you need to run with GTK <2.18, you also need to build with GTK <2.18.

If it's not this, then I don't understand what is going on.

**Comment by nieder** [comment]
I only have 1 copy of GTK+2 on this machine. It's 2.18.9. It does not define _gtk_widget_get_mapped since that didn't appear until 2.20.

I also don't understand why `#define gtk_widget_get_mapped(widget) (GTK_WIDGET_MAPPED(widget))` doesn't work when set in `plugins/gtkui/support.h`, but does in `plugins/gtkui/ddb_splitter.c` basically immediately after `support.h` gets #included.

Anyway, the patches here (plus the extra #define in ddb_splitter.c) did allow me to build deadbeef now. Thanks.

**Comment by Alexey-Yakovenko** [comment]
oh yes! I somehow confused it to be available from 2.18 and up. sorry for that. should be fixed now.
